### PR TITLE
⚠️ clusterctlv2: add the clusterctl init command

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds the `clusterctl init` command and completes the internal test framework that will be used by all the other commands

**Which issue(s) this PR fixes** 
Rif #1729

/assign @vincepri 
/cc @detiber @ncdc

### CODE REVIEW

File: cmd/clusterctl/cmd/init.go [hunk 1]
```diff
@@ -92,5 +95,27 @@ func init() {
...
 
 func runInit() error {
+	c, err := client.New(cfgFile)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("performing init...")
+
+	ComponentsList, firstExecution, err := c.Init(io.kubeconfig, io.coreProvider, io.bootstrapProviders, io.infrastructureProviders, io.targetNamespace, io.watchingNamespace, io.force)
```
Comment (vincepri) [inline cmd/clusterctl/cmd/init.go]:
```suggestion
	componentList, firstExecution, err := c.Init(io.kubeconfig, io.coreProvider, io.bootstrapProviders, io.infrastructureProviders, io.targetNamespace, io.watchingNamespace, io.force)
```

File: cmd/clusterctl/cmd/init.go [hunk 2]
```diff
@@ -92,5 +95,27 @@ func init() {
...
+	if err != nil {
+		return err
+	}
+
+	for _, components := range ComponentsList {
+		fmt.Printf(" - %s %s installed (%s)\n", components.Name(), components.Type(), components.Version())
+	}
+
+	if firstExecution {
+		fmt.Printf("\nYour cluster API management cluster has been initialized successfully!\n")
```
Comment (vincepri) [inline cmd/clusterctl/cmd/init.go]: Nit: Use `Println` if you don't need formatting

File: cmd/clusterctl/pkg/client/client_test.go [hunk 3]
```diff
@@ -33,41 +37,146 @@ func TestNewFakeClient(t *testing.T) {
 		WithVar("var", "value").
 		WithProvider(repository1Config)
 
-	// create a new fakeClient using the fake config
-	newFakeClient(config1)
+	// create a fake repository with some YAML files in it (usually matching the lit of providers defined in the config)
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/client_test.go]:
```suggestion
	// create a fake repository with some YAML files in it (usually matching the ? of providers defined in the config)
```

File: cmd/clusterctl/pkg/client/cluster/installer.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	// Perform the installation of the providers ready in the install queue.
+	Install() ([]repository.Components, error)
+}
+
+// installerService implements ProviderInstallerService
+type installerService struct {
+	proxy              Proxy
+	providerComponents ComponentsClient
+	providerInstaller  InventoryClient
+
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/installer.go]: extra space?

File: cmd/clusterctl/pkg/client/cluster/inventory.go [hunk 5]
```diff
@@ -42,8 +43,13 @@ type InventoryClient interface {
 	// Create an inventory item for a provider instance installed in the cluster.
 	Create(clusterctlv1.Provider) error
 
-	// Return inventory items for all the provider instances installed in the cluster.
+	// List the inventory items for all the provider instances installed in the cluster.
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/cluster/inventory.go]:
```suggestion
	// List returns the inventory items for all the provider instances installed in the cluster.
```

Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/inventory.go]: done

File: cmd/clusterctl/pkg/client/cluster/installer.go (hunk 4 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/installer.go]: done

**Comment by fabriziopandini** [comment]
@vincepri all comments addressed

File: cmd/clusterctl/pkg/client/cluster/installer.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+func (i *installerService) Install() ([]repository.Components, error) {
+	ret := make([]repository.Components, len(i.installQueue))
+	for c, components := range i.installQueue {
+		klog.V(3).Infof("Installing provider %s/%s:%s", components.TargetNamespace(), components.Name(), components.Version())
+
+		// create the provider
+		err := i.providerComponents.Create(components)
+		if err != nil {
```
Comment (Arvinderpal) [inline cmd/clusterctl/pkg/client/cluster/installer.go]: How do we plan to handle clean up in case something goes wrong? A simple rollback scheme should be sufficient IMO.

File: cmd/clusterctl/pkg/client/cluster/inventory.go [hunk 7]
```diff
@@ -215,3 +221,26 @@ func (p *inventoryClient) list(options listOptions) ([]clusterctlv1.Provider, er
...
+	for _, p := range l {
+		names.Insert(p.Name)
+	}
+
+	// if there is only one provider, this is the default
+	if names.Len() == 1 {
+		return names.List()[0], nil
+	}
+
+	// there no provider name/more than one provider name of this type; in both cases, it is not possible to get a default provider name
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/inventory.go]: fixed (hopefully now is more clear)

File: cmd/clusterctl/pkg/client/cluster/installer.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+	// Perform the installation of the providers ready in the install queue.
+	Install() ([]repository.Components, error)
+}
+
+// installerService implements ProviderInstallerService
+type installerService struct {
+	proxy              Proxy
+	providerComponents ComponentsClient
+	providerInstaller  InventoryClient
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/installer.go]: `providerInstaller  InventoryClient` was a mistake. It should be `providerInventory  InventoryClient` (fixed)

WRT to the naming of variables, my major concern ATM is to keep things consistent in the clusterctl codebase/while completing all the missing functionalities, so I'm keeping a note in the tracking issue and I will reconsider later

File: cmd/clusterctl/pkg/client/cluster/installer.go (hunk 6 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/installer.go]: what do you mean with a "simple rollback scheme"? 
From my test, the simple create or update logic implemented in Create are ok to cover most of the cases, so my current orientation is to not over-engineer in this first iteration...

**Comment by fabriziopandini** [comment]
@Arvinderpal comment addressed

File: cmd/clusterctl/pkg/client/cluster/installer.go (hunk 6 above)
Comment (Arvinderpal) [reply cmd/clusterctl/pkg/client/cluster/installer.go]: If `i.providerComponents.Create(components)` were throw an error in the middle -- for example, bootstrap components installed successfully but infra components install failed. There will be some "dangling" state left behind. 
We can leave it to the user to manually cleanup that state or provide some means to undo what has been installed. 
I realize this will complicate matters and I'm fine with it not being included.

Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/installer.go]: Once we will have `clusterctl delete` implemented, we can consider to automatically trigger delete if something goes wrong

File: cmd/clusterctl/pkg/client/common.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/repository"
+)
+
+// getComponentsByName is a utility methods that returns provider component for a given provider, targetNamespace and WatchingNamespace.
+func (c *clusterctlClient) getComponentsByName(provider string, targetNameSpace string, watchingNamespace string) (repository.Components, error) {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/common.go]:
```suggestion
func (c *clusterctlClient) getComponentsByName(provider string, targetNamespace string, watchingNamespace string) (repository.Components, error) {
```

File: cmd/clusterctl/pkg/client/init.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+import (
+	"github.com/pkg/errors"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/cluster"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+// Init a management cluster by adding the requested list of providers.
+
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/init.go]: Remove extra space

File: cmd/clusterctl/pkg/client/init.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+	// if not we consider this the first time init is executed, and thus we enforce the installation of a core provider (if not already explicitly requested by the user)
+	firstRun := false
+	currentCoreProvider, err := cluster.ProviderInventory().GetDefaultProviderName(clusterctlv1.CoreProviderType)
+	if err != nil {
+		return nil, false, err
+	}
+	if currentCoreProvider == "" && coreProvider == "" {
+		firstRun = true
+		coreProvider = config.ClusterAPIName
+	}
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/init.go]: I'm wondering if we could make this into its own method and remove the need to return more than 2 parameters

File: cmd/clusterctl/pkg/client/init.go [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+import (
+	"github.com/pkg/errors"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/cluster"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+// Init a management cluster by adding the requested list of providers.
+
+func (c *clusterctlClient) Init(kubeconfig string, coreProvider string, bootstrapProviders, infrastructureProviders []string, targetNameSpace, watchingNamespace string, force bool) ([]Components, bool, error) {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/init.go]: The parameter list seems rather large, should we use a struct instead?

File: cmd/clusterctl/pkg/client/init.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	// Components is an alias for repository.Components; this makes conversion
+	rr := make([]Components, len(r))
+	for i, components := range r {
+		rr[i] = components
+	}
+	return rr, firstRun, nil
+}
+
+// addToInstaller adds the components to the install queue and checks that the actual provider type match the target group
+func (c *clusterctlClient) addToInstaller(installer cluster.ProviderInstallerService, targetGroup clusterctlv1.ProviderType, targetNameSpace string, watchingNamespace string, force bool, providers ...string) error {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/init.go]: I'd also make the input here a struct

File: cmd/clusterctl/pkg/client/init.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+	if currentCoreProvider == "" && coreProvider == "" {
+		firstRun = true
+		coreProvider = config.ClusterAPIName
+	}
+
+	// create and installer service, add the requested providers to the install queue (thus performing validation of the target state of the management cluster
+	// before starting the installation), and then perform the installation.
+	installer := cluster.ProviderInstaller()
+
+	if coreProvider != "" {
```
Comment (vincepri) [inline cmd/clusterctl/pkg/client/init.go]: In what case the `coreProvider` isn't defined?

File: cmd/clusterctl/pkg/client/init.go (hunk 12 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init.go]: done

File: cmd/clusterctl/pkg/client/init.go (hunk 13 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init.go]: done

File: cmd/clusterctl/pkg/client/init.go (hunk 11 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init.go]: We can do in a follow-up PR when there will be a client for ProviderInventory() as well
I'm taking note in the tracking issue ATM

File: cmd/clusterctl/pkg/client/client.go [hunk 15]
```diff
@@ -17,25 +17,38 @@ limitations under the License.
...
+	configClient            config.Client
+	repositoryClientFactory RepositoryClientFactory
+	clusterClientFactory    ClusterClientFactory
 }
 
+type RepositoryClientFactory func(config.Provider) (repository.Client, error)
+type ClusterClientFactory func(string) (cluster.Client, error)
+
 // ensure clusterctlClient implements Client.
 var _ Client = &clusterctlClient{}
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/client.go]: I have this test across all the clusterctl codebase, not only here; I will make a separated PR to get all fixed.
I'm taking note in the tracking issue

File: cmd/clusterctl/pkg/client/init.go (hunk 14 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init.go]: Init can be executed multiple times for adding more providers to a management cluster

**Comment by fabriziopandini** [comment]
@vincepri rebased to fix the test error

File: cmd/clusterctl/pkg/client/client.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply cmd/clusterctl/pkg/client/client.go]: I normally put these in normal code, next to the type/struct definition. It's probably a stylistic choice and I'm not sure it's worth changing unless we plan to start codifying rules such as this and sticking to them during PR reviews.

File: cmd/clusterctl/pkg/client/client_test.go [hunk 16]
```diff
@@ -33,41 +37,146 @@ func TestNewFakeClient(t *testing.T) {
...
+
+func (f *fakeClient) WithRepository(repositoryClient repository.Client) *fakeClient {
+	if fc, ok := f.configClient.(fakeConfigClient); ok {
+		fc.WithProvider(repositoryClient)
+	}
+	f.repositories[repositoryClient.Name()] = repositoryClient
+	return f
+}
+
+// fakeClusterClient returns a cluster.Client for a fake management cluster that
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/client_test.go]: godoc is for a different method or type

File: cmd/clusterctl/pkg/client/cluster/client.go [hunk 17]
```diff
@@ -50,6 +50,10 @@ type Client interface {
 	// ProviderObjects returns a ObjectsClient object that can be user for
 	// operating cluster API objects stored in the management cluster (e.g. clusters, AWS clusters, machines etc.).
 	ProviderObjects() ObjectsClient
+
+	// ProviderInstaller returns an ProviderInstallerService that enforces consistency rules for provider installation,
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/client.go]:
```suggestion
	// ProviderInstaller returns a ProviderInstallerService that enforces consistency rules for provider installation,
```

File: cmd/clusterctl/pkg/client/cluster/installer.go [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+package cluster
+
+import (
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/repository"
+)
+
+// ProviderInstallerService defined methods for enforcing consistency rules for provider installation.
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/installer.go]:
```suggestion
// ProviderInstallerService defines methods for enforcing consistency rules for provider installation.
```

File: cmd/clusterctl/pkg/client/cluster/installer.go [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+type ProviderInstallerService interface {
+	// Add a provider to the install queue.
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/installer.go]:
```suggestion
	// Add adds a provider to the install queue.
```

File: cmd/clusterctl/pkg/client/cluster/installer.go [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+	// NB. By deferring the installation, the installer service can perform validation of the target state of the management cluster
+	// before actually starting the installation of new providers.
+	Add(repository.Components, bool) error
+
+	// Perform the installation of the providers ready in the install queue.
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/installer.go]:
```suggestion
	// Install performs the installation of the providers ready in the install queue.
```

File: cmd/clusterctl/pkg/client/cluster/installer.go [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+package cluster
+
+import (
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/repository"
+)
+
+// ProviderInstallerService defined methods for enforcing consistency rules for provider installation.
+type ProviderInstallerService interface {
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/installer.go]: Would it be possible to shorten this to just ProviderInstaller?

File: cmd/clusterctl/pkg/client/cluster/installer.go [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	i.installQueue = append(i.installQueue, components)
+	return nil
+}
+
+func (i *installerService) Install() ([]repository.Components, error) {
+	ret := make([]repository.Components, len(i.installQueue))
+	for c, components := range i.installQueue {
+		klog.V(3).Infof("Installing provider %s/%s:%s", components.TargetNamespace(), components.Name(), components.Version())
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/cluster/installer.go]: I believe we're trying to avoid package-level `klog` calls as much as possible. You don't need to change it here and now, but maybe add a todo in a tracker to switch to passing loggers around?

File: cmd/clusterctl/pkg/client/common.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/repository"
+)
+
+// getComponentsByName is a utility methods that returns provider component for a given provider, targetNamespace and WatchingNamespace.
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common.go]:
```suggestion
// getComponentsByName is a utility method that returns components for a given provider, targetNamespace, and watchingNamespace.
```

File: cmd/clusterctl/pkg/client/common.go [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+
+	// parse the abbreviated syntax for name[:version]
+	name, version, err := parseProviderName(provider)
+	if err != nil {
+		return nil, err
+	}
+
+	// if a target namespace is defined, use it, otherwise let the namespace empty so the default one defined in the
+	// provider components YAML will be used
+	var namespace string
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common.go]: I'm not sure you need this logic. I think you can just pass `targetNamespace` in on line 58 - if it's the empty string, it will use the default from the yaml; if it's not the empty string, it will use the value from the variable. Make sense?

File: cmd/clusterctl/pkg/client/common.go [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+	if targetNamespace != "" {
+		namespace = targetNamespace
+	}
+
+	// gets the provider configuration (that includes the location of the provider repository)
+	providerConfig, err := c.configClient.Providers().Get(name)
+	if err != nil {
+		return nil, err
+	}
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common.go]: Do you want to add a nil-check for providerConfig here, just to be safe?

File: cmd/clusterctl/pkg/client/common.go [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+// parseProviderName defines a utility function that parses the abbreviated syntax for name[:version]
+func parseProviderName(provider string) (name string, version string, err error) {
+	t := strings.Split(strings.ToLower(provider), ":")
+	if len(t) > 2 {
+		return "", "", errors.Errorf("invalid provider name %q. Provider name should be in the form [namespace/]name[:version]", provider)
+	}
+
+	name = t[0]
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common.go]: I would check that `t[0] != ""`

File: cmd/clusterctl/pkg/client/common.go [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+	errs := validation.IsDNS1123Label(name)
+	if len(errs) != 0 {
+		return "", "", errors.Errorf("invalid name value: %s", strings.Join(errs, "; "))
+	}
+
+	version = ""
+	if len(t) > 1 {
+		version = t[1]
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common.go]: I would check that `t[1] != ""`

File: cmd/clusterctl/pkg/client/common_test.go [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+import "testing"
+
+func Test_parseProviderName(t *testing.T) {
+	type args struct {
+		provider string
+	}
+	tests := []struct {
+		name          string
+		args          args
+		wantNamespace string
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common_test.go]: Unused?

File: cmd/clusterctl/pkg/client/common_test.go [hunk 29]
```diff
@@ -0,0 +1,69 @@
...
+			wantName:      "provider",
+			wantVersion:   "version",
+			wantErr:       false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gotName, gotVersion, err := parseProviderName(tt.args.provider)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("parseProviderName() error = %v, wantErr %v", err, tt.wantErr)
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common_test.go]: If you want, you can `t.Fatalf` here and omit the `return` on the next line

File: cmd/clusterctl/pkg/client/common_test.go [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+				return
+			}
+			if gotName != tt.wantName {
+				t.Errorf("parseProviderName() gotName = %v, want %v", gotName, tt.wantName)
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/common_test.go]: nit: specifying `parseProviderName()` here and in the other error lines is redundant with the test function name

File: cmd/clusterctl/pkg/client/init.go [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+package client
+
+import (
+	"github.com/pkg/errors"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/cluster"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+// Init a management cluster by adding the requested list of providers.
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init.go]:
```suggestion
// Init initializes a management cluster by adding the requested list of providers.
```

File: cmd/clusterctl/pkg/client/init.go [hunk 32]
```diff
@@ -0,0 +1,113 @@
...
+	currentCoreProvider, err := cluster.ProviderInventory().GetDefaultProviderName(clusterctlv1.CoreProviderType)
+	if err != nil {
+		return nil, false, err
+	}
+	if currentCoreProvider == "" && options.CoreProvider == "" {
+		firstRun = true
+		options.CoreProvider = config.ClusterAPIName
+	}
+
+	// create and installer service, add the requested providers to the install queue (thus performing validation of the target state of the management cluster
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init.go]:
```suggestion
	// create an installer service, add the requested providers to the install queue (thus performing validation of the target state of the management cluster
```

File: cmd/clusterctl/pkg/client/init.go [hunk 33]
```diff
@@ -0,0 +1,113 @@
...
+
+	// checks if the cluster already contains a Core provider.
+	// if not we consider this the first time init is executed, and thus we enforce the installation of a core provider (if not already explicitly requested by the user)
+	firstRun := false
+	currentCoreProvider, err := cluster.ProviderInventory().GetDefaultProviderName(clusterctlv1.CoreProviderType)
+	if err != nil {
+		return nil, false, err
+	}
+	if currentCoreProvider == "" && options.CoreProvider == "" {
+		firstRun = true
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init.go]: Why do we only set firstRun to true if options.CoreProvider is unset?

File: cmd/clusterctl/pkg/client/init.go [hunk 34]
```diff
@@ -0,0 +1,113 @@
...
+	if err := c.addToInstaller(addOptions, clusterctlv1.InfrastructureProviderType, options.InfrastructureProviders...); err != nil {
+		return nil, false, err
+	}
+
+	r, err := installer.Install()
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Components is an alias for repository.Components; this makes conversion
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init.go]: Incomplete comment

File: cmd/clusterctl/pkg/client/init.go [hunk 35]
```diff
@@ -0,0 +1,113 @@
...
+	rr := make([]Components, len(r))
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init.go]: I think I'd vote for more descriptive variable names than `r` and `rr`

File: cmd/clusterctl/pkg/client/init_test.go [hunk 36]
```diff
@@ -0,0 +1,379 @@
...
+				CoreProvider:            tt.args.coreProvider,
+				BootstrapProviders:      tt.args.bootstrapProvider,
+				InfrastructureProviders: tt.args.infrastructureProvider,
+				TargetNameSpace:         tt.args.targetNameSpace,
+				WatchingNamespace:       tt.args.watchingNamespace,
+				Force:                   tt.args.force,
+			})
+
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Init() error = %v, wantErr %v", err, tt.wantErr)
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init_test.go]: Can Fatalf and omit return

File: cmd/clusterctl/pkg/client/init_test.go [hunk 37]
```diff
@@ -0,0 +1,379 @@
...
+				return
+			}
+
+			if len(got) != len(tt.want) {
+				t.Errorf("Init() got = %v items, want %v items", len(got), len(tt.want))
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init_test.go]: Can Fatalf and omit return

File: cmd/clusterctl/pkg/client/init_test.go [hunk 38]
```diff
@@ -0,0 +1,379 @@
...
+				return
+			}
+
+			for i, g := range got {
+				w := tt.want[i]
+
+				if g.Name() != w.provider.Name() {
+					t.Errorf("Init(), Item[%d].Name() got = %v, want = %v ", i, g.Name(), w.provider.Name())
```
Comment (ncdc) [inline cmd/clusterctl/pkg/client/init_test.go]: Here & elsewhere: no need to include `Init()` in the error message

File: cmd/clusterctl/pkg/client/cluster/installer.go (hunk 21 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/installer.go]: done

File: cmd/clusterctl/pkg/client/cluster/installer.go (hunk 22 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/cluster/installer.go]: Ok! added to tracking issue

File: cmd/clusterctl/pkg/client/common.go (hunk 24 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/common.go]: yes! done

File: cmd/clusterctl/pkg/client/common.go (hunk 25 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/common.go]: Get return an object if the selected provider exists, or not in case it does not exists.
 This makes the nil check not necessary

File: cmd/clusterctl/pkg/client/common.go (hunk 26 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/common.go]: done

File: cmd/clusterctl/pkg/client/common.go (hunk 27 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/common.go]: done

File: cmd/clusterctl/pkg/client/common_test.go (hunk 28 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/common_test.go]: removed

File: cmd/clusterctl/pkg/client/common_test.go (hunk 29 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/common_test.go]: done

File: cmd/clusterctl/pkg/client/common_test.go (hunk 30 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/common_test.go]: done

File: cmd/clusterctl/pkg/client/init.go (hunk 33 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init.go]: Fixed. It is considered a first run if there is not yet a core provider; in this case, we add CAPI as a bootstrap provider only if the user di not specified something from the command line (e.g. a specific version of CAPI)

File: cmd/clusterctl/pkg/client/init.go (hunk 34 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init.go]: fixed

File: cmd/clusterctl/pkg/client/init.go (hunk 35 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init.go]: fixed

File: cmd/clusterctl/pkg/client/init_test.go (hunk 36 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/client/init_test.go]: fixed

**Comment by fabriziopandini** [comment]
@ncdc thanks for the feedback!
everything should be fixed now
